src/Akka.Streams.Kafka/Stages/CommittableConsumerStage.cs

```diff
@@ -32,51 +33,56 @@ public override ILogicAndMaterializedValue<Task> CreateLogicAndMaterializedValue
         private readonly ISubscription _subscription;
-        private readonly Outlet<CommittableMessage<K, V>> _out;
         private IConsumer<K, V> _consumer;
 
-        private Action<ConsumeResult<K, V>> _messagesReceived;
         private Action<IEnumerable<TopicPartition>> _partitionsAssigned;
         private Action<IEnumerable<TopicPartitionOffset>> _partitionsRevoked;
         private readonly Decider _decider;
 
-        private const string TimerKey = "PollTimer";
-
-        private readonly Queue<CommittableMessage<K, V>> _buffer;
         private IEnumerable<TopicPartition> _assignedPartitions;
-        private volatile bool _isPaused;
         private readonly TaskCompletionSource<NotUsed> _completion;
+        private readonly CancellationTokenSource _cancellationTokenSource;
 
         public KafkaCommittableSourceStage(CommittableConsumerStage<K, V> stage, Attributes attributes, TaskCompletionSource<NotUsed> completion) : base(stage.Shape)
         {
             _settings = stage.Settings;
             _subscription = stage.Subscription;
-            _out = stage.Out;
             _completion = completion;
-            _buffer = new Queue<CommittableMessage<K, V>>(stage.Settings.BufferSize);
+            _cancellationTokenSource = new CancellationTokenSource();
 
             var supervisionStrategy = attributes.GetAttribute<ActorAttributes.SupervisionStrategy>(null);
             _decider = supervisionStrategy != null ? supervisionStrategy.Decider : Deciders.ResumingDecider;
 
-            SetHandler(_out, onPull: () =>
+            SetHandler(stage.Out, onPull: () =>
             {
-                if (_buffer.Count > 0)
-                {
-                    Push(_out, _buffer.Dequeue());
-                }
-                else
+                try
                 {
-                    if (_isPaused)
+                    var message = _consumer.Consume(_cancellationTokenSource.Token);
+                    if (message == null) // No message received, or consume error occured
+                        return;
+
+                    var consumer = _consumer;
+                    var commitableOffset = new CommitableOffset(
+                        () => consumer.Commit(),
+                        new PartitionOffset("groupId", message.Topic, message.Partition, message.Offset));
+
+                    if (IsAvailable(stage.Out))
                     {
-                        _consumer.Resume(_assignedPartitions);
-                        _isPaused = false;
-                        Log.Debug("Polling resumed, buffer is empty");
+                        var commitableMessage = new CommittableMessage<K, V>(message, commitableOffset);
+                        Push(stage.Out, commitableMessage);
                     }
-                    PullQueue();
+                }
+                catch (OperationCanceledException)
+                {
+                    // Consume was canceled, looks like we are shutting down the stage
```

Comment:
Is there something we need to do to cancel the stage here? Or is there already going to be a stage termination in progress by the time this exception is thrown? And is `OperationCanceledException` thrown when the `CancellationToken` is invoked or is that going to throw a `TaskCancellationException` ?

Reply:
I am using it to handle `CancellationToken` usage, it was throwing `OperationCanceledException` when cancellation was requested. This cancellation is possible only when we get consume error and `HandleConsumeError` handler is invoked (currently on separate thread my the way). But may be this is because usually `HandleConsumeError` is called before next onPull is called and at the moment we call `_consumer.Consume(_cancellationTokenSource.Token)` token is already canceled and we get `OperationCanceledException`. And if for some reason  `_consumer.Consume(_cancellationTokenSource.Token)` will be called before cancellation is requested, we will get `TaskCancellationException`, because the consuming task will be already started internally... I will check this out and make update in separate PR. Possibly catching `TaskCancellationException` as well will be enough.

Reply:
We'll see how things go during integration testing